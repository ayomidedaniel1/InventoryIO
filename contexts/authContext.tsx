import React, {
  useState,
  useEffect,
  useContext,
  createContext,
  FC,
  ReactNode,
} from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

export type User = {
  email: string;
};

type AuthContextType = {
  user: User | null;
  isLoggedIn: boolean;
  loading: boolean;
  login: (user: User) => void;
  logout: () => void;
};

type AuthProviderProps = {
  children: ReactNode;
};

export const AuthContext = createContext<AuthContextType>(
  undefined as never
);

export const AuthProvider: FC<AuthProviderProps> = ({ children }) => {
  const [user, setUser] = useState<User | null>(null);
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    checkUser();
  }, []);

  const checkUser = async () => {
    setLoading(true);
    try {
      const isUserLoggedIn = await AsyncStorage.getItem('isLoggedIn');
      const storedUser = await AsyncStorage.getItem('user');

      if (isUserLoggedIn !== null && storedUser) {
        setIsLoggedIn(true);
        setUser(JSON.parse(storedUser));
      }
    } catch (error) {
      console.error("Error retrieving user ", error);
    } finally {
      setLoading(false);
    }
  };

  const login = (newUser: User) => {
    AsyncStorage.setItem("isLoggedIn", 'true');
    AsyncStorage.setItem("user", JSON.stringify(newUser));
    setIsLoggedIn(true);
    setUser(newUser);
  };

  const logout = () => {
    setUser(null);
    AsyncStorage.removeItem("isLoggedIn");
    AsyncStorage.removeItem("user");
    setIsLoggedIn(false);
  };

  const contextValue: AuthContextType = {
    user,
    loading,
    isLoggedIn,
    login,
    logout,
  };

  return (
    <AuthContext.Provider value={contextValue}>
      {children}
    </AuthContext.Provider>
  );
};

export function useAuthContext(): AuthContextType {
  return useContext(AuthContext);
}