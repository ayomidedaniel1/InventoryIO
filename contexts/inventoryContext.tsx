import React, {
  useState,
  useEffect,
  useContext,
  createContext,
  FC,
  ReactNode,
} from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { useAuthContext } from "./authContext";

type InventoryItem = {
  name: string;
  totalStock: number;
  price: number;
  description: string;
};

type InventoryContextType = {
  inventory: InventoryItem[];
  addInventory: (item: InventoryItem) => void;
  editInventory: (name: string, newItem: InventoryItem) => void;
  deleteInventory: (name: string) => void;
};

type InventoryProviderProps = {
  children: ReactNode;
};

export const InventoryContext = createContext<InventoryContextType | null>(null);

export const InventoryProvider: FC<InventoryProviderProps> = ({ children }) => {
  const authContext = useAuthContext();
  const [inventory, setInventory] = useState<InventoryItem[]>([]);

  useEffect(() => {
    const loadInventory = async () => {
      try {
        const storedInventory = await AsyncStorage.getItem('inventory');

        if (storedInventory) {
          setInventory(JSON.parse(storedInventory));
        }
      } catch (error) {
        console.error("Error loading inventory ", error);
      }
    };

    loadInventory();
  }, []);

  const addInventory = async (item: InventoryItem) => {
    const newInventory = [...inventory, item];
    setInventory(newInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(newInventory));
  };

  const editInventory = async (name: string, newItem: InventoryItem) => {
    const tempInventory = [...inventory];

    const updatedInventory = tempInventory.map(item => {
      return item.name === name ? newItem : item;
    });
    setInventory(updatedInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(updatedInventory));
  };

  const deleteInventory = async (name: string) => {
    const updatedInventory = inventory.filter(item => item.name !== name);
    setInventory(updatedInventory);
    await AsyncStorage.setItem('inventory', JSON.stringify(updatedInventory));
  };

  if (!authContext.isLoggedIn) {
    return null;
  }

  const contextValue: InventoryContextType = {
    inventory,
    addInventory,
    editInventory,
    deleteInventory,
  };

  return (
    <InventoryContext.Provider value={contextValue}>
      {children}
    </InventoryContext.Provider>
  );
};

export function useInventoryContext(): InventoryContextType {
  const context = useContext(InventoryContext);
  if (!context) {
    throw new Error("Some error occurred");
  }
  return context;
}