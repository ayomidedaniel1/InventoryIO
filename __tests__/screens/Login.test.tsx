import React from "react";
import renderer from 'react-test-renderer';
import { AuthProvider } from "../../contexts/authContext";

jest.mock('react-native-simple-toast', () => ({
  show: jest.fn(),
}));

import LoginScreen from "../../app/(auth)/login";

test('LoginScreen renders correctly', () => {
  const tree = renderer.create(
    <AuthProvider>
      <LoginScreen />
    </AuthProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});