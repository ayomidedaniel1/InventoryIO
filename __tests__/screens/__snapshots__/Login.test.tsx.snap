// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "column",
      "paddingHorizontal": 20,
      "paddingTop": 50,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#6c47ff",
        "height": 80,
        "marginBottom": 50,
        "marginTop": -50,
        "width": 750,
      }
    }
  />
  <ViewManagerAdapter_ExpoImage
    contentFit="cover"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={147}
    marginBottom={20}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={[]}
    source={[]}
    style={
      {
        "height": 147,
        "marginBottom": 20,
        "width": "100%",
      }
    }
    transition={
      {
        "duration": 1000,
      }
    }
    width="100%"
  />
  <View
    style={
      {
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 21,
          "marginVertical": 10,
          "textAlign": "center",
        }
      }
    >
      Welcome
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 16,
          "lineHeight": 29,
          "textAlign": "center",
        }
      }
    >
      By signing in you agree to our
    </Text>
    <Text
      style={
        {
          "color": "#0386D0",
          "fontSize": 16,
          "lineHeight": 29,
          "marginBottom": 12,
          "textAlign": "center",
        }
      }
    >
      Term and privacy policy
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F9F9F9",
          "borderColor": "#CCC9C9",
          "borderRadius": 9,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 47,
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 20,
          "paddingLeft": 20,
          "paddingRight": 40,
        }
      }
    >
      <TextInput
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Email Address"
        placeholderTextColor="#C4C4C4"
        selectionColor="#CCC9C9"
        style={
          {
            "color": "#000",
            "fontSize": 15,
            "height": "100%",
            "width": "100%",
          }
        }
      />
      <Text />
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F9F9F9",
          "borderColor": "#CCC9C9",
          "borderRadius": 9,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 47,
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 20,
          "paddingLeft": 20,
          "paddingRight": 40,
        }
      }
    >
      <TextInput
        autoComplete="password"
        keyboardType="default"
        onChangeText={[Function]}
        placeholder="Password"
        placeholderTextColor="#C4C4C4"
        returnKeyType="send"
        secureTextEntry={true}
        selectionColor="#CCC9C9"
        style={
          {
            "color": "#000",
            "fontSize": 15,
            "height": "100%",
            "width": "100%",
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#0386D0",
          "borderRadius": 27,
          "height": 45,
          "justifyContent": "center",
          "marginTop": 52,
          "opacity": 1,
          "width": 151,
        }
      }
    >
      <Text
        style={
          {
            "color": "#FFF",
            "fontSize": 14,
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
